hare/hare.go

```diff
@@ -368,6 +370,13 @@ func (h *Hare) onTick(ctx context.Context, lid types.LayerID) (bool, error) {
 		nonce = &nnc
 	}
 
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if len(h.cps) >= h.config.LimitConcurrent {
+		err := fmt.Errorf("failed to start cp for layer %q hare reached concurrent cp limit %q", lid.Uint32(), h.config.LimitConcurrent)
+		logger.With().Warning("hare failure", log.Err(err))
```

Comment:
log the details (layer and limit) here and only propagate the error with minimal content.
errors are usually wrapped many levels by the time it propagates up. too long of error msg will clogged up upstream logging.

Reply:
The problem I see with returning simple errors is that you lose the context of the error, so when it reaches the top level and you get the error `hare reached concurrent limit` you're in the dark as to what actually happened. Logs can help but if you have multiple processes or threads you may be unable to determine what caused the error you are looking at. Also for testing we often don't have logs enabled so it can be really helpful in determining why a test fails when the error contains all the context.

I'm not saying that there couldn't be a problem with upstream logging being clogged, but I think there is a balance to be met between adding contextual information to returned errors and deciding what to log.

Just having a blanket policy of returning simple errors and logging details can lead to logs that don't let you understand what actually went wrong.

In this case the result of `onTick` is ignored except in tests so I don't see a longer error message here being a problem.

Reply:
your argument seems to contradicts itself...

if you want more context for debugging, then you should log the context instead of packing it in the error object, since the caller of onTick() ignore the error.

for me there is no hard rule in how much you log and pack in the error. the responsibility of logging should rest on the entity that handles errors. if the upstream caller handles the error, then i support packing the context in the error itself. but here it's not the case.

Reply:
What's different here as opposed to [this comment](https://github.com/spacemeshos/go-spacemesh/pull/4130#discussion_r1128497406), where I thought you were saying that because the caller can change it's good practice to include all the context in the error irrespective of what the current caller is doing with it?

Reply:
like i explained. the diff is who gets to log it.

if you are logging it anyway, then you log the context and/or pack the context in the error
if you are passing the error upstream and not log it, then you pack the context in the error.

this discussion seems longer than it needs to be. lets add the contexts (layer/limit) in the log message also. i don't really care (at this point) what's in the error.